### PR TITLE
i#2971 rm CI: Cleanup: Remove CLIENT_INTERFACE define

### PR DESCRIPTION
**Pull request by derekbruening**
Removes the CLIENT_INTERFACE define.  It is now always enabled.
As part of this, removes all of the IF_CLIENT_INTERFACE, etc. macros.

Issue: #2971

### CODE REVIEW

**Comment by derekbruening** [comment]
The diff is enormous.  Not expecting a reviewer to comb through every line -- too time-intensive.  Maybe a couple spot-checks. 
 Most of it was created with automation, although there were some errors in some steps there.  But I think we have to mostly rely on our tests here.

**Comment by derekbruening** [comment]
Friendly ping (again, nobody is expecting manual checking of every line for this type of diff -- though if you really wanted to do that and are in the middle of it please provide a status).

File: core/lib/instrument_api.h [hunk 1]
```diff
@@ -6548,14 +6541,11 @@ DR_API
...
 dr_trace_exists_at(void *drcontext, void *tag);
-#    endif /* CUSTOM_TRACES */
-
-#endif /* CLIENT_INTERFACE */
+#endif /* CUSTOM_TRACES */
 
 /****************************************************************************
  * proc.c routines exported here due to proc.h being in arch_exports.h
  * which is included in places where opnd_t isn't a complete type.
- * These are used for dr_insert_clean_call() and thus are not just CLIENT_INTERFACE.
```
Comment (sapostolakis) [inline core/lib/instrument_api.h:6558]: Maybe keep this part: "These are used for dr_insert_clean_call()."

**Review by sapostolakis — APPROVED** [review]

**Comment by sapostolakis** [comment]
Sorry for the delay. Looked into the whole thing and it required a couple of sittings. Only have a few nit changes to recommend.

File: core/heap.c [hunk 2]
```diff
@@ -260,8 +258,7 @@ typedef struct _thread_units_t {
  * should require couldbelinking status, which makes them safe wrt unlink flushing.
  * Xref DrMi#1791.
  */
-#define SEPARATE_NONPERSISTENT_HEAP() \
-    (DYNAMO_OPTION(enable_reset) IF_CLIENT_INTERFACE(|| true))
+#define SEPARATE_NONPERSISTENT_HEAP() (DYNAMO_OPTION(enable_reset) || true)
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply core/heap.c:261]: Removed, and moved the comment to the global_nonpersistent_units field.

File: core/globals.h [hunk 3]
```diff
@@ -95,18 +95,6 @@
...
-/* we only export IR interface if CLIENT_INTERFACE is defined */
-#ifdef CLIENT_INTERFACE
-/* in Makefile, we define these (so that genapi.pl, etc. get them):
- *    define DYNAMORIO_IR_EXPORTS 1
- *    define CUSTOM_TRACES 1
- * CUSTOM_TRACES_RET_REMOVAL is aggressive -- assumes calling convention kept
- * Only useful if custom traces are doing inlining => do not define for external
- * release, or even by default for internal since it's always on even if
- * not building custom traces!
- */
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply core/globals.h:107]: It's hard to find it in this large diff, but the useful content here, the CUSTOM_TRACES_RET_REMOVAL comment, was moved to configure.cmake.h.

File: core/dispatch.c [hunk 4]
```diff
@@ -1199,12 +1194,12 @@ dispatch_exit_fcache(dcontext_t *dcontext)
          * todo list so we should never get here
          */
         if (SHARED_FRAGMENTS_ENABLED()) {
-            USAGE_ERROR("CLIENT_INTERFACE incompatible with -shared_{bbs,traces}"
+            USAGE_ERROR("client to_do incompatible with -shared_{bbs,traces}"
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply core/dispatch.c:1197]: -shared_{bbs,traces} is on by default.  The original was nonsensical.  This to_do list is used only for thread-private-cache-only API routines.  But the error message will not make sense to a user who will have no idea what to_do is. Rewriting to mention the two routines: dr_{delete,replace}_fragment.

File: core/arch/interp.c [hunk 5]
```diff
@@ -4599,7 +4568,8 @@ decode_trace(void *drcontext, void *tag)
             d_r_mutex_lock(&thread_initexit_lock);
         ilist = recreate_fragment_ilist(dcontext, NULL, &frag, &alloc_res,
                                         false /*no mangling*/
-                                        _IF_CLIENT(false /*do not re-call client*/));
+                                        ,
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply core/arch/interp.c:4571]: Yeah clang-format doesn't seem willing to pull it across the comment or next to it.  I already tweaked a bunch of these pre-posting.  Searching again: found a couple more like this one and fixed them.

File: core/translate.c [hunk 6]
```diff
@@ -1923,9 +1890,9 @@ stress_test_recreate_state(dcontext_t *dcontext, fragment_t *f, instrlist_t *ili
          * regions): not ideal to test using part of what we're testing but
          * better than nothing
          */
-        ilist = recreate_fragment_ilist(dcontext, NULL, &f, NULL,
-                                        true /*mangle*/
-                                        _IF_CLIENT(true /*call client*/));
+        ilist = recreate_fragment_ilist(dcontext, NULL, &f, NULL, true /*mangle*/
+                                        ,
+                                        true /*call client*/);
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply core/translate.c:1895]: Already found as part of handling last comment.

File: core/optionsx.h [hunk 7]
```diff
@@ -826,18 +820,15 @@ OPTION_DEFAULT(uint_time, deadlock_timeout,
 
 /* stack_size may be adjusted by adjust_defaults_for_page_size(). */
 OPTION_DEFAULT(uint_size, stack_size,
-               /* the CI build has a larger MAX_OPTIONS_STRING so we need
+               /* For client we have a larger MAX_OPTIONS_STRING so we need
                 * a larger stack even w/ no client present.
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply core/optionsx.h:824]: You mean sthg like add an 's"?  s/client/clients/.

File: core/lib/instrument_api.h (hunk 1 above)
Comment (derekbruening) [reply core/lib/instrument_api.h:6558]: Done.

**Comment by derekbruening** [comment]
> Sorry for the delay. Looked into the whole thing and it required a couple of sittings. Only have a few nit changes to recommend.

Thank you for the thorough review!  It was good to have another set of eyes on the whole thing.

**Pull request merged**